core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -657,6 +688,43 @@ class BlockManagerMasterEndpoint(
     }
   }
 
+  private def getShufflePushMergerLocations(
+      numMergersNeeded: Int,
+      hostsToFilter: Set[String]): Seq[BlockManagerId] = {
+    val blockManagersWithExecutors = blockManagerIdByExecutor.groupBy(_._2.host)
+      .mapValues(_.head).values.map(_._2).toSet
+    val filteredBlockManagersWithExecutors = blockManagersWithExecutors
+      .filterNot(x => hostsToFilter.contains(x.host))
+    val filteredMergersWithExecutors = filteredBlockManagersWithExecutors.map(
+      x => BlockManagerId(x.executorId, x.host, StorageUtils.externalShuffleServicePort(conf)))
+
+    // Enough mergers are available as part of active executors list
+    if (filteredMergersWithExecutors.size >= numMergersNeeded) {
+      filteredMergersWithExecutors.toSeq
+    } else {
+      // Delta mergers added from inactive mergers list to the active mergers list
+      val filteredMergersWithExecutorsHosts = filteredMergersWithExecutors.map(_.host)
+      val filteredMergersWithoutExecutors = shuffleMergerLocations.values
+        .filterNot(x => hostsToFilter.contains(x.host))
+        .filterNot(x => filteredMergersWithExecutorsHosts.contains(x.host))
+      val randomFilteredMergersLocations =
+        if (filteredMergersWithoutExecutors.size >
+          numMergersNeeded - filteredMergersWithExecutors.size) {
+          Utils.randomize(filteredMergersWithoutExecutors)
+        } else {
+          filteredMergersWithoutExecutors
+        }
+      filteredMergersWithExecutors.toSeq ++ randomFilteredMergersLocations
+        .take(numMergersNeeded - filteredMergersWithExecutors.size)
```

Comment:
We can only perform `take()` when `randomize()` is performed.

Reply:
If # elements in `randomFilteredMergersLocations` is less than `numMergersNeeded - filteredMergersWithExecutors.size`, `take()` would just return all elements.
What's the reason for `take()` cannot be called here if `randomize()` is not performed?

Reply:
@Ngone51 Is it because `filteredMergersWithoutExecutors` is an `Iterable`?

Reply:
I checked the `Iterable` size and its not scanning through the collection. It seems to work fine. Also we have a test case which is covering this case. Let me know if I'm missing something here. Thanks